File: 백준/Gold/7579. 앱/앱.py
import sys
input= sys.stdin.readline

N,M=map(int,input().split())

Mem=[0]+list(map(int,input().split()))
C=[0]+list(map(int,input().split()))
ans=sys.maxsize

DP=[[0 for _ in range(sum(C)+1)] for _ in range(N+1)]

for i in range(1,N+1):
    byte=Mem[i]
    cost=C[i]
    for j in range(1,sum(C)+1):
        #현재 앱을 비활성화 시킬만큼, cost가 충분하지 않은 경우
        if j<cost:
            DP[i][j]=DP[i-1][j]
        else:
            #현재 코스트가 충분한 경우는 , 2가지 경우가 있을수 있다.
            DP[i][j]= max(DP[i-1][j],DP[i-1][j-cost]+byte)
        if DP[i][j]>=M:
            ans=min(ans,j)

print(ans)

File: 백준/Gold/7579. 앱/앱.cc

#include <iostream>
#include <vector> 
#include <string> 
#include <cstring> 
#include <algorithm> 
#include <cmath> 
#include <unordered_map>
#include <map> 
#include <queue> 
#include <set> 
#define jl "\n"
using namespace std; 
int N , M ; 
int maxv = pow(2,31) -1 ; 


int main() { 

    cin >> N >> M ; 

    int dp[N+1][10001] ;
    memset(dp, 0, sizeof(dp));
    int memory[N+1] ;
    int cost[N+1] ; 
    int ans = maxv ; 
    for(int i=1 ; i<= N; i++){
        cin >> memory[i]  ; 
    }
    for(int i=1 ; i<= N; i++){
        cin >> cost[i]  ; 
    }

    for(int x=1 ; x <=N ; x++){
        for(int y=1 ; y <= 10000 ; y++){

            if( y  >= cost[x] ){
                dp[x][y] = max( dp[x-1][y] , dp[x-1][y-cost[x]] + memory[x] ) ;
            } else {
                dp[x][y] = dp[x-1][y] ;
            }
            if(M<= dp[x][y]){
                ans = min(ans , y) ; 
            }
        }
    } cout << ans ;


}


